### ISSUE TITLE
MySQL table compiler charset usage incompatible with npm mysql API

### ISSUE THREAD

**Issue by emilgoldsmith** [tgriesser/knex#2742]
Sorry, ditching the template because I think I can explain it clearly enough and I'm going to submit a PR asap with a possible fix.

Basically as can be seen here https://github.com/mysqljs/mysql/blob/cf5d1e396a343ffa0fba23b3791d2dd5da20f30e/Readme.md#connection-options the documentation for the `charset` config option in `mysql` is (and has been so the last 4 years): 

> `charset`: The charset for the connection. This is called "collation" in the SQL-level of MySQL (like `utf8_general_ci`). If a SQL-level charset is specified (like `utf8mb4`) then the default collation for that charset is used. (Default: `'UTF8_GENERAL_CI'`)

But in the Knex table compiler code there's the following https://github.com/tgriesser/knex/blob/26dfeb8805bb11b63d76065d244c78c7b380b41b/src/dialects/mysql/schema/tablecompiler.js#L34-L41

which sets what may either be a charset (that mysql finds the default collation for) or a collation as the charset. This therefore fails all my migrations when I changed to using `utf8mb4_unicode_ci` as my `charset`. Also, as far as I can see `mysql` has never accepted a `collate` option, so I'm not sure where that comes from? 

I have a fix in mind and will work on it right now and submit a PR asap, I just thought it would be good form to submit an issue first, then I can also link to this for the reasoning for my PR.

**Comment by mercmobily** [comment]
@emilgoldsmith Do you happen to be still around?

**Comment by mercmobily** [comment]
Thanks for identifying this charset/collation handling issue with the MySQL table compiler.

You're absolutely right - there's a mismatch between how the mysql driver handles the `charset` option and how Knex's table compiler uses it. The mysql driver's `charset` option accepts either a charset (like `utf8mb4`) or a collation (like `utf8mb4_unicode_ci`), and when given a collation, it uses that directly. But Knex's table compiler incorrectly assumes `charset` is always a character set.

Looking at the current code in lib/dialects/mysql/schema/mysql-tablecompiler.js:40-45:

```javascript
const charset = this.single.charset || conn.charset || '';
const collation = this.single.collate || conn.collate || '';
const engine = this.single.engine || '';

if (charset && !like) sql += ` default character set ${charset}`;
if (collation) sql += ` collate ${collation}`;
```

This code treats `conn.charset` as a character set, but the mysql driver documentation clearly states that `charset` can be either a charset or a collation. When you pass `utf8mb4_unicode_ci` as the charset in your connection config, Knex incorrectly generates:

```sql
CREATE TABLE ... default character set utf8mb4_unicode_ci
```

Which is invalid SQL - `utf8mb4_unicode_ci` is a collation, not a charset.

The table compiler needs to detect whether the provided `charset` value is actually a collation. If it's a collation (identifiable by containing an underscore after the charset name, like `utf8mb4_unicode_ci`), it should be used with the `COLLATE` clause instead.

A possible approach would be:
1. Check if the charset value looks like a collation (contains underscore pattern)
2. If it's a collation, use it with `COLLATE` clause
3. If it's a charset, use it with `DEFAULT CHARACTER SET` clause
4. Maintain backward compatibility for existing code

Regarding the `collate` option you mentioned - you're correct that the mysql driver doesn't accept a `collate` option in its connection settings. The `collate` property in Knex's table compiler appears to be for table-level specification during table creation, not a connection option.

**Workaround**

Until this is fixed, you can work around the issue by:
1. Using just the charset part (`utf8mb4`) in your connection config and specifying collation at table creation
2. Or explicitly setting collation when creating tables using Knex's schema builder

**Related issues**

- #2666 - Fix collation when renaming column in MySQL dialect
- #5568 - Add collate to the columnbuilder interface
- PR #2752 - Your WIP fix for this issue (still open)

**Comment by mercmobily** [comment]
After thorough analysis of this issue and PR #2752, I believe we should **close this PR and start fresh** with a much simpler solution.

**Why close the PR #2752**

1. The approach of introspecting driver internals (`require('mysql/lib/protocol/constants/charsets')`) is fragile and will break when drivers reorganise their internal structure
2. The maintainer feedback from @elhigu about architectural separation between mysql and mysql2 dialects is valid but adds complexity for this problem
3. The PR has been stalled since 2018 with unresolved architectural concerns

**The simpler solution:**

MySQL has a consistent, well-documented naming convention that has been stable for 20+ years:
- **Charsets** never contain underscores: `utf8mb4`, `latin1`, `binary`
- **Collations** always contain underscores: `utf8mb4_unicode_ci`, `latin1_swedish_ci`

We can solve this with a simple pattern check:

```javascript
// In mysql-tablecompiler.js
parseCharsetValue(value) {
  if (!value || typeof value !== 'string') return { charset: '', collation: '' };
  
  const normalized = value.toUpperCase();
  
  // If contains underscore, it's a collation
  if (normalized.includes('_')) {
    // Extract charset from first part before underscore
    const charset = normalized.split('_')[0];
    return {
      charset: charset,
      collation: normalized
    };
  }
  
  // Otherwise it's a charset
  return {
    charset: normalized,
    collation: ''
  };
}

// Then in createTable()
const connectionCharset = conn.charset || '';
const { charset: parsedCharset, collation: parsedCollation } = this.parseCharsetValue(connectionCharset);

const charset = this.single.charset || parsedCharset || '';
const collation = this.single.collate || parsedCollation || '';

if (charset && !like) sql += ` default character set ${charset}`;
if (collation) sql += ` collate ${collation}`;
```

**Benefits of this approach:**

1. **No external dependencies** - relies only on MySQL's naming convention
2. **Works identically for mysql and mysql2** - no dialect-specific code needed
3. **Simple and maintainable** - ~15 lines of code vs 167 in the current PR
4. **Future-proof** - MySQL's naming convention is unlikely to change
5. **Backward compatible** - handles both charset and collation values correctly

**Testing:**
- `charset: 'utf8mb4'` → `DEFAULT CHARACTER SET UTF8MB4`
- `charset: 'utf8mb4_unicode_ci'` → `DEFAULT CHARACTER SET UTF8MB4 COLLATE UTF8MB4_UNICODE_CI`
- Existing table-level charset/collate options continue to work as before

**Recommendation:**

Let's close PR #2752 with thanks to @emilgoldsmith for identifying the issue and initial work. I can submit a fresh, minimal PR implementing the pattern-matching approach above, which will be much easier to review, test, and maintain.

This issue has affected users for 6+ years - let's finally resolve it with the simplest solution that works.

**Comment by emilgoldsmith** [comment]
> @emilgoldsmith Do you happen to be still around?

I am haha, and since my student days when I wrote this I have also graduated to senior engineer now haha. 

Don't have a ton of bandwidth to dive into this anymore and not an active user of knex at the moment either I'm afraid.

Anything I can help with though?

**Comment by mercmobily** [comment]
@emilgoldsmith eheheh This is what we get for waiting 7 years to answer an issue!!!
I mean, now that the problem is described clearly in this thread, I would loooOOooove a PR. But if you are not an active user, I recognise that this would be just a distraction for you, so all good 😄

**Comment by emilgoldsmith** [comment]
Yeah afraid I'm going to have to skip that this time around! But feel free to close my issues (or i can do it if you dont have permissions)

**Comment by mercmobily** [comment]
Oh no the issues are still valid all good. we will fix them for everybody else regardless
